### DISCUSSION TITLE
Language support for Zip-like simultaneous enumeration

### DISCUSSION THREAD

**Discussion by Richiban** [dotnet/csharplang#2197] [Language Ideas] [answered]
I propose we support (what I hope is) a very simple addition to `foreach`: allowing the enumeration of multiple sequences simultaneously, in much the same fashion as the concept of zipping.

Syntax:

```csharp
foreach(var a in collectionA; var b in collectionB) 
{
	// This block will execute only while both `a` and `b` continue to yield values

	// Code goes here
}
```

This would lower to:

```csharp
var $enumeratorA = collectionA.GetEnumerator();
var $enumeratorB = collectionB.GetEnumerator();

while($enumeratorA.MoveNext() && $enumeratorB.MoveNext())
{
	var a =$ enumeratorA.Current;
	var b = $enumeratorB.Current;

	// Code goes here
}
```

I find myself writing code like this a lot:

```csharp
foreach(var (a, b) in collectionA.Zip(collectionB, (x, y) => (x, y)))
{

}
```

This would also (indirectly) provide a feature that is very oft-requested: looping over a collection with an automatic counter variable. If we combine this with ranges then I think we'll be able to write:

```csharp
foreach(var x in col; var i in 0..)
{

}
```

**Comment by quinmars** [comment]
I think it'd be better to ask the corefx team to add a default value tuple selector or to write that extension method on your own. I think:

```csharp
foreach (var (a, b) in collectionA.Zip(collectionB))
{

}
```

is very readable. So you don't need any language support.

**Comment by Logerfo** [comment]
Here's the implementation I've been using:
```c#
/// <exception cref="ArgumentNullException"><paramref name="first"/> or <paramref name="second"/> is <see langword="null"/></exception>
public static IEnumerable<(TFirst first, TSecond second)> Zip<TFirst, TSecond>(this IEnumerable<TFirst> first, IEnumerable<TSecond> second)
   => (first ?? throw new ArgumentNullException(nameof(first))).Zip(second ?? throw new ArgumentNullException(nameof(second)), (i1, i2) => (i1, i2));
```
The language feature would be nice, tho.

**Comment by jnm2** [comment]
See https://github.com/dotnet/corefx/issues/16011.

**Comment by YairHalberstadt** [comment, marked as answer]
Using extension GetEnumerator from C# 9.0, you can do this:

```csharp
foreach(var (a, b) in (collectionA, collectionB)) 
{
	// This block will execute only while both `a` and `b` continue to yield values

	// Code goes here
}

public static IEnumerator<(T1, T2)> GetEnumerator<T1, T2>(this (IEnumerable<T1> a, IEnumerable<T2> b) pair) => pair.a.Zip(pair.b, (a, b) => (a, b)).GetEnumerator(); 
```

**Reply by Joe4evr** [reply]
~~You don't even need extension `GetEnumerator()` because [an overload of `Zip()` that yields tuples was introduced in Core 3.0](https://apisof.net/catalog/System.Linq.Enumerable.Zip%3CTFirst,TSecond%3E(IEnumerable%3CTFirst%3E,IEnumerable%3CTSecond%3E))~~
Whoops, I see this is doing something slightly different. #cantread

**Reply by Logerfo** [reply]
Could this extensions be implemented on the framework itself with a few arities?

**Reply by YairHalberstadt** [reply]
Feel free to open a request on dotnet/runtime
